### DISCUSSION TITLE
Best match for ExceptionShielding in CoreWCF

### DISCUSSION THREAD

**Discussion by oskarjf** [CoreWCF/CoreWCF#1212] [Q&A] [answered]
Hi
I'm upgrading a WCF project to CoreWCF and I'm not sure what to do with ExceptionShielding as that is not available in CoreWCF.
Is there a possibility for a general exception handler?

**Comment by mconnew** [comment]
What is ExceptionShielding? It's not a part of WCF. The exception handling api's are the same other than the namespace so recompiling code using the different namespace should "just work" in most cases.

**Reply by oskarjf** [reply]
It's part of Microsoft Enterprise Library. Just a way of catching all exceptions and handle them.
[Documentation](https://learn.microsoft.com/en-us/previous-versions/msp-n-p/ff648319(v=pandp.10)?redirectedfrom=MSDN)

**Reply by g7ed6e** [reply, marked as answer]
Please have a look to WCF error handling feature https://learn.microsoft.com/en-us/dotnet/api/system.servicemodel.dispatcher.ierrorhandler?view=dotnet-plat-ext-7.0
You should be able to handle exceptions implementing an `IErrorHandler`.

**Reply by mconnew** [reply]
I don't know anything about the Microsoft Enterprise Library implementation, what it can do for you or if it's worth bringing over, but I found [the source code here](https://github.com/thalesfu/EnterpriseLibrary6/tree/master/Source/Blocks/ExceptionHandling/Src/WCF). I can't vouch for if this is the genuine source code and that it hasn't been altered, so use at your own risk. It should be easy enough to port over as all the type names are the same and nearly all apis are the exact same shape with the only variations being a small number are now async (I don't think any are in this area).  
If you want to try porting it, I'll answer any questions that you might have if there's an ambiguity on how to port some of the code.

**Reply by igor-nesterov-deltatre** [reply]
@oskarjf  thanks for follow up:)
My similar question was ignored:
https://github.com/CoreWCF/CoreWCF/discussions/1167
I guess one of the "killer feature" of the MSEntLib 
that you can specify exception handling in wcf.config as I described in my question.
@oskarjf  I guess it can be easily ported by yourself, since CoreWCF is not planning to do that
@g7ed6e  does not seem like a replacement, but for know there is no alternatives

**Reply by mconnew** [reply]
@igor-nesterov-deltatre, I'm sorry we missed your question. 
I wouldn't be opposed to shipping these features as part of CoreWCF. I personally don't have the bandwidth to do the port, I already have too many pieces up in the air that I need to complete. I think the right place would be to create a CoreWCF.EnterpriseLibrary package containing a ported implementation (and potentially any other appropriate EnterpriseLibrary features). Obviously the built-in mechanism in WCF/CoreWCF is too bare bones otherwise this wouldn't exist. Improving on the error handling capability natively in CoreWCF is a good idea, and new feature work to do so is likely to look different than the EnterpriseLibrary solution. I'd rather not have multiple improvements (and multiple alternative api's) in the Primitives package as it makes it confusing as to what's the recommended approach. I'd rather have EnterpriseLibrary be a compat porting library.  
If anyone wants to contribute porting the code, please let me know.

**Reply by igor-nesterov-deltatre** [reply]
@mconnew  I will happy  to do that, once I will have any results and ensure, that it's possible at least, I will drop the message here

**Reply by mconnew** [reply]
If you hit any problems, let me know. It should port over successfully as I've kept the api shapes the same.

**Reply by igor-nesterov-deltatre** [reply]
hi @oskarjf . 
I find out that it heavily depends from xml configuration configs,  which are not implemented in CoreWCF.
Am I right that it was done for the purpose and now preferable way to customize everything in code or using .json configs.
I can try reimplement xml classes for CoreWCF, but I guess pr will not be accepted,

**Reply by mconnew** [reply]
We do have CoreWCF.ConfigurationManager which has support for the traditional app.config. My future plan for configuration is create a way to load service config from the newer `IConfiguration` mechanism and convert CoreWCF.ConfigurationManager into as `IConfigurationProvider` implementation which provides the data in the `IConfiguration` way of doing things based on the app.config.  
I don't know how easy it would be to add what you need to CoreWCF.ConfigurationManager. It shouldn't be a requirement as that package currently pulls in everything else, so a code based approach would be needed too. Can you point me at how configuration is currently being used and I might be able to come up with a clean design. The complication with CoreWCF.ConfigurationManager was all the WCF stuff we have in machine.config which is necessary for everything to work. If there isn't a dependency on the machine.config stuff, then it might be possible to do something standalone.

**Reply by mconnew** [reply]
I just did a deep dive on the code base and its documentation and I think supporting app.config will be fine. There's a small design change you'll need, but not much. The library has a concept of an `IConfigurationSource` which is where the configuration is loaded from. By default it pulls it from your app.config/web.config file in .NET Framework. There's a utility class called `FileConfigurationSource` which allows you to have a .config file named anything you want. My suggestion is that an app must add an `IConfigurationSource` to DI and this library will pull that from DI to load the config. Have an extension method on `IServiceCollection` to add EnterpriseServices support which has a parameter that takes the config file name. Something like this:
```c#
public IServiceCollection AddEnterpriseServices(this IServiceCollection services, string filepath)
{
    if (!FileExists(filepath)) throw new Exception(); // Do something better, but you get the gist
    services.AddSingleton(IConfigurationSource, new FileConfigurationSource(filepath));
}
```
You will likely end up needing to add other stuff too, but the idea is `IConfigurationSource` is added to DI with a source which reads from a named config file. Normally on .NET Framework you would add a config entry in your main app.config file which specifies the configuration source. This should be the only piece that config will need to move to code. Everything else you should be able to leave in an app.config file. The documentation for how you would have configured the config source in your main app.config file [can be found here](https://documentation.help/MS-Enterprise-Library-5.0/EntLib50_d294bdc2-2c9d-4358-96ac-bff4017afd0f.html).
